pkg/api/v1beta3/types.go

```diff
@@ -934,7 +934,7 @@ type ServiceSpec struct {
 	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
 
 	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
-	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
+	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
```

Comment:
Do we *have* to reduce the grammar in the client tools to the combined form?  I worry about the approachability of the query selector a bit to client authors.  It's *extremely* valuable for the client to be able to correlate Service -> RC -> pod.  If we used the expanded form (simple conditions split) then client tools would know that the selector is *at least* that selective, but the edge cases they could ignore (i.e. the "" query can't add things, only remove them).

----- Original Message -----
> > @@ -934,7 +934,7 @@ type ServiceSpec struct {
> >  	Ports []ServicePort `json:"ports" description:"ports exposed by the
> >  	service"`
> >  
> >  	// This service will route traffic to pods having labels matching this
> >  	selector. If null, no endpoints will be automatically created. If empty,
> >  	all pods will be selected.
> > -	Selector map[string]string `json:"selector" description:"label keys and
> > values that must match in order to receive traffic for this service; if
> > empty, all pods are selected, if not specified, endpoints must be manually
> > specified"`
> > +	Selector string `json:"selector" description:"label keys and values that
> > must match in order to receive traffic for this service; if empty, all
> > pods are selected, if not specified, endpoints must be manually
> > specified"`
> 
> I see your point but client authors will be impacted in any cases. as soon
> we'll change the grammar in the value of the map entry with the empty string
> key (`{"": "foo, foo not in (baz)"}`).
> 
> I agree that the powerful queries will be rare.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7053/files#r28876688